src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,228 @@
+                        currentChild.setSuffixLink(parentSuffix.getChild().get(cc));
+                    } else {
+                        currentChild.setSuffixLink(root);
+                    }
+
+                    q.add(currentChild); // Add the child node to the queue for further processing
+                }
+
+                // Establish output links for nodes to efficiently identify patterns within patterns
+                if (currentState.getSuffixLink().getPatternInd() >= 0) {
+                    currentState.setOutputLink(currentState.getSuffixLink());
+                } else {
+                    currentState.setOutputLink(currentState.getSuffixLink().getOutputLink());
+                }
+            }
+        }
+
+        // Searches for patterns in the input text and records their positions
+        public ArrayList<ArrayList<Integer>> searchIn(String text) {
+            Node parent = root; // Start searching from the root node
+            for (int i = 0; i < text.length(); i++) {
+                char ch = text.charAt(i); // Get the current character in the text
+
+                // Check if the current node has a child for the current character
+                if (parent.getChild().containsKey(ch)) {
+                    parent = parent.getChild().get(ch); // Update the current node to the child node
+
+                    // If the current node represents a pattern, record its position in res
+                    if (parent.getPatternInd() > -1) {
+                        res.get(parent.getPatternInd()).add(i);
+                    }
+
+                    Node outputLink = parent.getOutputLink();
+                    // Follow output links to find and record positions of other patterns
+                    while (outputLink != null) {
+                        res.get(outputLink.getPatternInd()).add(i);
+                        outputLink = outputLink.getOutputLink();
+                    }
+                } else {
+                    // If no child node exists for the character, backtrack using suffix links
+                    while (parent != root && !parent.getChild().containsKey(ch)) {
+                        parent = parent.getSuffixLink();
+                    }
+                    if (parent.getChild().containsKey(ch)) {
+                        i--; // Decrement i to reprocess the same character
+                    }
+                }
+            }
+
+            /*
+             * By default res contains end-points.
+             * This function converts those endpoints to start points
+             */
+            for (int i = 0; i < patterns.length; i++) {
+                if (!res.get(i).isEmpty()) {
+                    for (int j = 0; j < res.get(i).size(); j++) {
+                        int endpoint = res.get(i).get(j);
+                        res.get(i).set(j, endpoint - patterns[i].length() + 1);
+                    }
+                }
+            }

```

Comment:
This could go to a separate `private` method.

Reply:
same reason as above for why i did it, but no problem, I will revert it back into separate method.